service-pdf/tests/OpgTest/Lpa/Pdf/AbstractPdfTestClass.php

```diff
@@ -0,0 +1,249 @@
+    {
+        $config = Config::getInstance();
+        return $config['service']['assets']['template_path_on_ram_disk'] . '/' . $templateName;
+    }
+
+    protected function verifyExpectedPdfData(AbstractIndividualPdf $pdf, $templateFileName, $strikeThroughTargets, $blankTargets, $constituentPdfs, $data, $pageShift, $formattedLpaRef = null)
+    {
+        //  Verify the provided expected data is as expected
+        $this->verifyTemplateFileName($pdf, $templateFileName);
+        $this->verifyStrikeThroughTargets($pdf, $strikeThroughTargets);
+        $this->verifyBlankTargets($pdf, $blankTargets);
+        $this->verifyConstituentPdfs($pdf, $constituentPdfs);
+        $this->verifyData($pdf, $data);
+        $this->verifyPageShift($pdf, $pageShift);
+
+        if (!is_null($formattedLpaRef)) {
+            $this->verifyFormattedLpaRef($pdf, $formattedLpaRef);
+        }
+    }
+
+    private function verifyFormattedLpaRef(AbstractIndividualPdf $pdf, $expectedValue)
+    {
+        $this->verifyReflectionProperty('formattedLpaRef', $pdf, $expectedValue);
+    }
+
+    private function verifyTemplateFileName(AbstractIndividualPdf $pdf, $expectedValue)
+    {
+        $this->verifyReflectionProperty('templateFileName', $pdf, $expectedValue);
+    }
+
+    private function verifyData(AbstractIndividualPdf $pdf, $expectedValue)
+    {
+        $this->verifyReflectionProperty('data', $pdf, $expectedValue);
+    }
+
+    private function verifyStrikeThroughTargets(AbstractIndividualPdf $pdf, $expectedValue)
+    {
+        $this->verifyReflectionProperty('strikeThroughTargets', $pdf, $expectedValue);
+    }
+
+    private function verifyBlankTargets(AbstractIndividualPdf $pdf, $expectedValue)
+    {
+        $this->verifyReflectionProperty('blankTargets', $pdf, $expectedValue);
+    }
+
+    private function verifyConstituentPdfs(AbstractIndividualPdf $pdf, $expectedValue)
+    {
+        //  First loop through the reflection values and swap out the instances where the pdf set is actually a PDF object
+        //  We do not need to test that here
+        $constituentPdfsForPages = $this->reflectionProperties['constituentPdfs']->getValue($pdf);
+
+        foreach ($constituentPdfsForPages as $page => $constituentPdfsForPage) {
+            foreach ($constituentPdfsForPage as $constituentPdfIdx => $constituentPdfForPage) {
+                //  Confirm that there is a corresponding entry in the expected values
+                if (!isset($expectedValue[$page]) || !array_key_exists($constituentPdfIdx, $expectedValue[$page])) {
+                    $this->fail(sprintf('Expected value missing while trying to verify constituent PDF index %s inserted after page %s', $constituentPdfIdx, $page));
+                }
+
+                $expectedPdf = $expectedValue[$page][$constituentPdfIdx];
+                $constituentPdf = $constituentPdfForPage['pdf'];
+
+                if ($constituentPdf instanceof AbstractIndividualPdf) {
```

Comment:
I always worry when I see instanceOf .... feels like when that happens, the functions maybe ought to be on separate classes , and use inheritance/polymorphism to ensure the right function gets called for the class. However, this may be difficult to disentangle now I guess, assuming this is existing test code. Once for another story perhaps?

Reply:
The problem (which I haven't fixed) is figuring out how to work out which PDFs are embedded in aggregator PDFs, which is bypassed here, effectively by ignoring them. In addition, some PDFs are included in the parent PDF as strings where the PDF doesn't need any data injected into it (e.g. a guidance page which is always the same). Rather than treat all constituent PDFs uniformly, we therefore have this mix of AbstractIndividualPdf objects, aggregator objects, and strings referencing the location of a PDF; which is why we therefore have this mess in the test.

Short of rewriting the whole thing to behave more sensibly and consistently, we have to keep this work-around to deal with that unsatisfactory data model.

Reply:
Fair enough :)